### PR TITLE
Support configuration cached builds

### PR DESCRIPTION
**Pull request by clayburn**
Fixes https://github.com/etiennestuder/teamcity-build-scan-plugin/issues/160

This PR modifies the init script to allow printing of the build scan url in scenarios where a configuration cache entry is reused. To handle the build started message, the logic is moved into the server components to react to TeamCity APIs for the start of the build.

### CODE REVIEW

**Comment by etiennestuder** [comment]
Thanks, @clayburn. These changes are very deep and it will take me a while to review them.

**Comment by etiennestuder** [comment]
Thanks, @clayburn. I've integrated the script related changes and the testing.

The BUILD_STARTED event is a performance optimization that was important back when log parsing was enabled by default. GIven there is not really any need for log parsing anymore due to both Gradle runner and Cmd line runner being instrumented with the init script, we'll simply drop supporting the performance optimization for the scenario where a Gradle build is run with config caching enabled AND log parsing enabled AND a huge log being parsed.

File: agent/src/main/resources/build-scan-init.gradle [hunk 1]
```diff
@@ -170,28 +168,39 @@ static def extensionsWithPublicType(def container, String publicType) {
...
+    BuildScanLifeCycleLogger(logger) {
+        this.logger = logger
     }
-}
 
-static boolean isNotAtLeast(String versionUnderTest, String referenceVersion) {
-    GradleVersion.version(versionUnderTest) < GradleVersion.version(referenceVersion)
-}
+    void log(String message) {
+        logger.log(generateBuildScanLifeCycleMessage(message))
```
Comment (etiennestuder) [inline agent/src/main/resources/build-scan-init.gradle:183]: Shouldn't this say `logger.quiet` ?

Comment (clayburn) [reply agent/src/main/resources/build-scan-init.gradle:183]: Yes, thanks for catching that

**Pull request closed without merging**